src/nplinker/arranger.py

```diff
@@ -0,0 +1,446 @@
+
+        The datasets include MIBiG, GNPS, antiSMASH, and BiG-SCAPE.
+        """
+        # The order of arranging the datasets matters, as some datasets depend on others
+        self.arrange_mibig()
+        self.arrange_gnps()
+        self.arrange_antismash()
+        self.arrange_bigscape()
+        self.arrange_strain_mappings()
+        self.arrange_strains_selected()
+
+    def arrange_podp_project_json(self) -> None:
+        """Arrange the PODP project JSON file.
+
+        If `config.mode` is "podp", download the PODP project JSON file if it doesn't exist. Then
+        validate the PODP project JSON file if it exists or is downloaded.
+
+        The validation is controlled by the json schema `schemas/podp_adapted_schema.json`.
+        """
+        if config.mode == "podp":
+            file_name = f"paired_datarecord_{config.podp_id}.json"
+            podp_file = globals.DOWNLOADS_DEFAULT_PATH / file_name
+            if not podp_file.exists():
+                download_url(
+                    PODP_PROJECT_URL.format(config.podp_id),
+                    globals.DOWNLOADS_DEFAULT_PATH,
+                    file_name,
+                )
+
+            with open(podp_file, "r") as f:
+                json_data = json.load(f)
+            validate_podp_json(json_data)
+
+    def arrange_mibig(self) -> None:
+        """Arrange the MIBiG metadata.
+
+        Always download and extract the MIBiG metadata if `config.mibig.to_use` is True.
+        If the default directory has already existed, it will be removed and re-downloaded to ensure
+        the latest version is used. So it's not allowed to manually put MIBiG metadata in the
+        default directory.
+        """
+        if config.mibig.to_use:
+            if globals.MIBIG_DEFAULT_PATH.exists():
+                # remove existing mibig data
+                shutil.rmtree(globals.MIBIG_DEFAULT_PATH)
+            download_and_extract_mibig_metadata(
+                globals.DOWNLOADS_DEFAULT_PATH,
+                globals.MIBIG_DEFAULT_PATH,
+                version=config.mibig.version,
+            )
+
+    def arrange_gnps(self) -> None:
+        """Arrange the GNPS data.
+
+        If `config.mode` is "local", validate the GNPS data directory.
+        If `config.mode` is "podp", download the GNPS data if it doesn't exist or remove the
+        existing GNPS data and re-download it if it is invalid.
+
+        The validation process includes:
+        - Check if the GNPS data directory exists.
+        - Check if the required files exist in the GNPS data directory, including:
+            - file_mappings.tsv or file_mappings.csv
+            - spectra.mgf
+            - molecular_families.tsv
+            - annotations.tsv
+        """
+        if config.mode == "local":
+            validate_gnps(globals.GNPS_DEFAULT_PATH)
+
+        if config.mode == "podp":
+            # set range 3 to ensure download can try 2 times and downloaded data is valid
```

Comment:
Good catch! It can loop at most 3 times. Multiple times are required to make sure the downloaded data is valid; if not, try to download it again (and again).  The downloaded data might be invalid in the case of e.g. losing connection to internet, nplinker process killed, etc.

I updated this part of code and related ones to fix a bug in the logics: after 3 times of loop, the data should be validated again. See the new commit please.